zerver/worker/queue_processors.py

```diff
@@ -607,17 +608,44 @@ def maybe_send_batched_emails(self) -> None:
 
 
 @assign_queue("email_senders")
-class EmailSendingWorker(QueueProcessingWorker):
-    @retry_send_email_failures
-    def consume(self, event: Dict[str, Any]) -> None:
-        # Copy the event, so that we don't pass the `failed_tries'
-        # data to send_email_from_dict (which neither takes that
-        # argument nor needs that data).
-        copied_event = copy.deepcopy(event)
-        if "failed_tries" in copied_event:
-            del copied_event["failed_tries"]
-        handle_send_email_format_changes(copied_event)
-        send_email_from_dict(copied_event)
+class EmailSendingWorker(LoopQueueProcessingWorker):
+    def consume_batch(self, events: List[Dict[str, Any]]) -> None:
+        @retry_send_email_failures
+        def send_email(worker: EmailSendingWorker, event: Dict[str, Any]) -> None:
+            send_email_from_dict(event)
+
+        try:
+            connection = initialize_connection()
+        except OSError as e:
+            self._handle_consume_exception(events, e)
+        else:
+            copied_events = copy.deepcopy(events)
+            for copied_event in copied_events:
+                # don't pass the `failed_tries' data to send_email_from_dict
+                # (which neither takes that argument nor needs that data).
+                if "failed_tries" in copied_event:
+                    del copied_event["failed_tries"]
+                copied_event["connection"] = connection
+                handle_send_email_format_changes(copied_event)
+                send_email(self, copied_event)
+            try:
+                connection.close()
```

Comment:
We explicitly _don't_ want to close the connection after every email we send.

Reply:
Indeed, my bad there. Instead I overwrote the stop() function of EmailSendingWorker to close the connection when we stop the queue.